far/vmenu.cpp

```diff
@@ -140,15 +140,45 @@ void VMenu::ResetCursor()
 	GetCursorType(PrevCursorVisible,PrevCursorSize);
 }
 
+template <typename coll, typename pred>
+int FindNearestSelectableItem(const coll& Coll, const int Pos, pred&& Pred, const bool GoBackward, const bool DoWrap)
+{
+	assert(0 <= Pos && Pos < static_cast<int>(Coll.size()));
```

Comment:
> 2. If we're asserting that Pos is non-negative, why not make it size_t?

The context where we call this function uses `int`s. We need to stitch 'int' and 'size_t` somewhere. I think the "right" place is on the boundary of this function. If you think otherwise, I can change it; I am not strong on this point.

> 1. If we're asserting that Pos is non-negative, it's better to cast Pos to size_t than size to int.

If you've meant "If we're asserting that Pos is _less than the size_", I'll change it.

Reply:
> If you think otherwise

In general, I think that function's signature should reflect its behavior.
FindNearestSelectableItem works with non-negative positions (and asserts that), so why not just reflect it.

> If you've meant "If we're asserting that Pos is less than the size", I'll change it.

I mean that there are 2 way to silence the warning:
1. make both sides int
2. make both sides size_t

<span>#</span>2 is not applicable when the number can be negative, for obvious reasons.
When the number cannot be negative, it is preferable to <span>#</span>1, because it won't potentially truncate.

Reply:
> I think that function's signature should reflect its behavior.

I think the function signature should reflect its contract (I am not sure about the word "behavior"). And the contract should generally be written for the convenience of the caller as opposed to the convenience of the implementation. If we take this view, the `Pos` here should be `int` because we are going to call this function with a signed `Pos` (why, it's a different story).

Now, if we change the signature to 
```
int FindNearestSelectableItem(const coll& Coll, const size_t Pos, pred&& Pred, const bool GoBackward, const bool DoWrap)
{
	assert(Pos < Coll.size());
	. . .
```
everything _happens to work._ However, it works only by a coincidence, because signed integers are stored in two's complement notation and binding a `size_t` argument to an `int` parameter does not generate a warning (probably for a good reason, but whatever). Strictly speaking, implicit conversion from `int` to `size_t` may overflow and yield an unexpected result on a weird platform.

Putting it the other way around, if we have a signed value and expect it to be within a `[0, size)` range, somebody should explicitly assert this value to be non-negative instead of depending on the implicit conversion which will turn negative values into "too large" positives. After all, on a 64-bit platform, we can legitimately have more than 2^31 items. However, if `Pos` happens to be equal to `-42`, it more likely is a bug which better be caught explicitly, even if in debug mode only.

Reply:
> does not generate a warning (probably for a good reason, but whatever)

Probably because such a conversion is way too common.
However, implicit is implicit, so tools might start complaining one day.

Fair enough, let's keep int.